### PR TITLE
Enhancement/338/improve logging

### PR DESCRIPTION
**Pull request by julianxyz**

### CODE REVIEW

File: packages/sakuli-core/src/bootstrap/load-bootstrap-options.function.ts [hunk 1]
```diff
@@ -28,11 +28,21 @@ export async function loadBootstrapOptions(): Promise<SakuliBootstrapOptions> {
...
     for (const packagePreset of packagePresets) {
       if (!sakuliBootstrapOptions.presetProvider.includes(packagePreset)) {
         sakuliBootstrapOptions.presetProvider.push(packagePreset);
       }
     }
+    if (sakuliBootstrapOptions.presetProvider.length) {
+      console.info(
+        `Loading Sakuli with ${sakuliBootstrapOptions.presetProvider}`
+      );
+    }
```
Comment (s1hofmann) [inline packages/sakuli-core/src/bootstrap/load-bootstrap-options.function.ts:45]: I'd love to have this info in a log file as well. Maybe in `packages/sakuli-core/src/sakuli.class.ts`

File: packages/sakuli-legacy/src/context/common/actions/action.function.spec.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+      expect(ctx.endTestAction).toHaveBeenCalled();
+  });
+
+  it("should log an error if action did throw an error", async () => {
+      withActionContext = runAsAction(ctx, "myAction", () => {throw Error("myError");});
+
+      await expect(withActionContext()).rejects.toThrow("Error in action: myAction \nmyError");
+
+      expect(ctx.logger.trace).toHaveBeenCalledTimes(2);
+      expect(ctx.endTestAction).toHaveBeenCalled();
```
Comment (s1hofmann) [inline packages/sakuli-legacy/src/context/common/actions/action.function.spec.ts:40]: I would expect a call to `ctx.startTestAction` as well.

File: packages/sakuli-legacy/src/context/common/actions/action.function.spec.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  beforeEach(() => {
+    ctx = createTestExecutionContextMock();
+  });
+
+  it("should execute successfully", async () => {
+      withActionContext = runAsAction(ctx, "myAction", () => "worked");
+      await withActionContext();
+
+      expect(ctx.logger.trace).toHaveBeenCalledTimes(2);
+      expect(ctx.endTestAction).toHaveBeenCalled();
```
Comment (s1hofmann) [inline packages/sakuli-legacy/src/context/common/actions/action.function.spec.ts:31]: I would expect a call to `ctx.startTestAction` as well.

**Comment by cliffle** [comment]
Won't the sakuli/sakuli-enterprise-forwarder#49 Issue gonna be resolved in #325? :sweat_smile:

**Comment by s1hofmann** [comment]
> Won't the [sakuli/sakuli-enterprise-forwarder#49](https://github.com/sakuli/sakuli-enterprise-forwarder/issues/49) Issue gonna be resolved in #325? 😅

Yes, but it could've been done here, too 😄

File: packages/sakuli-cli/src/cli-utils/test-execution-context-renderer.function.ts [hunk 4]
```diff
@@ -75,24 +84,63 @@ const renderEntityOnEnd = (
...
       .on("START_TESTSTEP", (s) => l(renderEntityOnStart(s, "Step", 3)))
       .on("END_TESTSTEP", (s) => {
         l(
           ansiEscapes.cursorUp(1) +
             ansiEscapes.eraseLine +
-            renderEntityOnEnd(s, "Step", 3)
+            ansiEscapes.cursorUp(1)
         );
+        logEntityOnEnd(s, "Step", 3);
+      })
```
[the remark this answers is not in the document]
Comment (s1hofmann) [reply packages/sakuli-cli/src/cli-utils/test-execution-context-renderer.function.ts]: I just noticed that when skipping the whole `START_TESTSTEP` event we're also skipping the log output.
So we might just skip the console render but keep the log file output

File: packages/sakuli-cli/src/cli-utils/test-execution-context-renderer.function.ts [hunk 5]
```diff
@@ -75,24 +81,57 @@ const renderEntityOnEnd = (
...
-      .on("START_TESTSTEP", (s) => l(renderEntityOnStart(s, "Step", 3)))
+      .on("START_EXECUTION", (_) => {
+        l(`Started execution`);
+      })
+      .on("START_TESTSUITE", (s) => {
+        logEntityOnStart(s, "Testsuite");
+      })
+      .on("START_TESTCASE", (s) => {
+        logEntityOnStart(s, "Testcase", 2);
+      })
```
Comment (s1hofmann) [inline packages/sakuli-cli/src/cli-utils/test-execution-context-renderer.function.ts:125]: `START_TESTSTEP` should be written to the log file

**Comment by s1hofmann** [comment]
We could keep the start scroll into view log by extending the `WebElement` mock

```
getAttribute: (attributeName: string) => Promise.resolve(`mockAttribute${attributeName}`),
getId: () => Promise.resolve("mockElementId"),
getTagName: () => Promise.resolve("mockTagName"),
getText: () => Promise.resolve("mockElementText"),
```

**Pull request merged**